lib/charms/grafana_agent/v0/cos_agent.py

```diff
@@ -586,6 +1021,55 @@ def trigger_refresh(self, _):
         # FIXME: Figure out what we should do here
         self.on.data_changed.emit()  # pyright: ignore
 
+    def _get_requested_protocols(self, relation: Relation):
+        # Coherence check
+        units = relation.units
+        if len(units) > 1:
+            # should never happen
+            raise ValueError(
+                f"unexpected error: subordinate relation {relation} "
+                f"should have exactly one unit"
+            )
+
+        unit = next(iter(units), None)
+
+        if not unit:
+            return None
+
+        if not (raw := relation.data[unit].get(CosAgentProviderUnitData.KEY)):
+            return None
+
+        if not (provider_data := self._validated_provider_data(raw)):
+            return None
+
+        return provider_data.tracing_protocols
+
+    def requested_protocols(self):
+        """All receiver protocols that have been requested by our related apps."""
+        requested_protocols = set()
+        for relation in self._charm.model.relations[self._relation_name]:
+            try:
+                protocols = self._get_requested_protocols(relation)
+            except NotReadyError:
+                continue
+            if protocols:
+                requested_protocols.update(protocols)
+        return requested_protocols
+
+    def _get_receiver_url(self, protocol: str):
+        if (
+            hasattr(self._charm, "cert")
+            and hasattr(self._charm.cert, "enabled")  # type: ignore
+            and self._charm.cert.enabled  # type: ignore
+        ):
+            s = "s"
+        else:
+            s = ""
```

Comment:
```suggestion
        scheme = "https" if self._charm.cert.available else "http"
```
See https://github.com/canonical/observability-libs/pull/96

Reply:
interesting, first I want to know why we're coding so defensively with all those hasattrs.
If we really need to do that, I'd rather have:
```
scheme = "http"
try:
    if self._charm.cert.enabled:
        scheme = "https"
except AttributeError:
    pass?    
```

Reply:
There was quite a lot of assumptions for the implementing charm here - that it both has the `cert` attribute and that `cert` attribute has an `enabled` method (which is true for grafana-agent but might not necessarily be true to any charm implementing cos-agent ever). That's why I went pretty defensive here and I think Pietro's approach might make sense. It also makes sense to write a proper comment here.

Reply:
ok, grafana-agent is using cert-handler v0 for juju 2.9 compatibility, so we don't have `enabled` here. Switched to `cert.available` as this is, well, available.